File: POH7/Mizugi/Main.scala
/* Try POH
 * author: Leonardone @ NEETSDKASU
 */
object Main extends App {
    
    def reject(v :Long, d :Long, c :Int) :(Long, Int) = {
        if (v % d == 0)
            reject(v / d, d, c + 1)
        else
            (v, c)
    }
    
    def conti(n :Long, r :Long, c :Int) :(Long, Int) = {
        if (n > 1) {
            val (x, c1) = reject(n, 5, 0)
            val (y, c2) = reject(x, 2, 0)
            val z = (r * y) % 1000000000
            conti(n - 1, z, c + c2 - c1)
        } else
            (r, c)
    }
    
    def multiple(v :Long, c :Int) :Long = {
        if (c == 0)
            v
        else
            multiple((v * 2) % 1000000000, c - 1)
    }
    
    def solve(n :Long) :Long = {
        val (r, c) = conti(n, 1, 0)
        multiple(r, c)
    }
    
    val n = scala.io.StdIn.readLine.toLong
    println(solve(n))
}

File: POH7/Mizugi/Main.fs
// Try POH
// author: Leonardone @ NEETSDKASU
(* --------------------------------------------------- *)
let (/%) v d = let dv = v / d in (dv, v - dv * d)
let flip f x y = f y x
let strsplit (str : string) = str.Split
let gs = stdin.ReadLine
let gi = int << gs
let gl = int64 << gs
let gss = gs >> flip strsplit [|' '|]
let gis = gss >> Array.map int
let ngt n f = [| for _ in 1 .. n -> f() |]
let ngs n = ngt n gs
let ngi n = ngt n gi
let ngss n = ngt n gss
let ngis n = ngt n gis
(* --------------------------------------------------- *)

let solve() =
    let modval v = v % 1000000000L
    let rec reject d v c =
        match v /% d with
        | (dv, 0L) ->  reject d dv (c + 1)
        | _ -> (v, c)
    let acc (r, c) v =
        let (x, c1) = reject 5L v 0
        let (y, c2) = reject 2L x -c1
        (modval <| r * y, c + c2)
    let n = gl()
    let (r, c) = seq{ 1L .. n } |> flip Seq.fold (1L, 0) acc
    let ans = Seq.fold (fun r _ -> modval <| r * 2L) r <| seq { 1 .. c }
    printfn "%d" ans


(* --------------------------------------------------- *)
[<EntryPoint>]
let main _ = solve(); 0

;;
